### PR TITLE
Chapter3 answers

### PR DESCRIPTION
**Pull request by aiwaverse**
### Solutions for Chapter 3

This took a VERY long time, and honestly stressed me more than I thought it would.
I'm aware my solutions are not perfect, I also made very liberal use of Either (I really like Either!), I hope this is ok.

In the end I feel my typeclasses are a little too big, specially because I found really hard to make both monsters and knights "work" together, I found them really different. Also the way I needed to use error is just ugly, but I failed to come up with a better solution.

Still, I really hope my solutions are enough for hacktoberfest, and onwards to the last one that I hope is easier :P 

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter3.hs [hunk 1]
```diff
@@ -50,8 +50,10 @@ signatures in places where you can't by default. We believe it's helpful to
 provide more top-level type signatures, especially when learning Haskell.
 -}
 {-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
```
Comment (chshersh) [inline src/Chapter3.hs]: Hehe, anticipating some interesting code after seeing this extension 😅 
But looks like it wasn't needed in the end...

File: src/Chapter3.hs [hunk 2]
```diff
@@ -376,6 +385,26 @@ after the fight. The battle has the following possible outcomes:
...
+               , knightGold   :: Int
+               }
+
+fight :: Monster -> Knight -> Int
+fight m k | newMonsterHealth <= 0 = knightGold k + monsterGold m
+          | newKnightHealth <= 0 = -1
+          | otherwise = knightGold k
+  where
+    newMonsterHealth = monsterHealth m - knightAttack k
+    newKnightHealth = knightHealth k - monsterAttack m
```
Comment (chshersh) [inline src/Chapter3.hs:405]: Great solution 👏🏻

File: src/Chapter3.hs [hunk 3]
```diff
@@ -482,6 +519,49 @@ After defining the city, implement the following functions:
...
 -}
 
+-- up to four people living in the house
+data House = One | Two | Three | Four deriving Show
+
+-- castle with a name and possibly a wall
+data Castle = Castle String Wall deriving Show
+
+-- the wall type, if true, wall was built
+newtype Wall = Wall Bool deriving Show
```
Comment (chshersh) [inline src/Chapter3.hs]: Your solution to this task is amazing 🔝 
Actually, it's one of the best I've seen!

I have only one suggestion but it's not relevant to requirements, it's more about style. Instead of having a `newtype` over a `Bool` (which is already great, you correctly noticed that `newtype` can be used here instead of `data`), it's probably better to implement a custom enum:

```haskell
data Wall = NoWalls | Wall
```

It can be slightly more readable to construct a new castle without walls like
`Castle newCastleName NoWalls` instead of existing `Castle newCastleName (Wall False)`.

Basically, it's a good general pattern to replace `Bool`s that have some specialized semantics with a custom data type. At least in Haskell, it's cheap to create new data types 🙂

File: src/Chapter3.hs [hunk 4]
```diff
@@ -482,6 +519,49 @@ After defining the city, implement the following functions:
...
+
+buildWalls c = c
+
+countPeople :: City -> Int
+countPeople City {houses = h} = foldl' (\a e -> a + countHouse e) 0 h
+  where
+    countHouse One = 1
+    countHouse Two = 2
+    countHouse Three = 3
+    countHouse Four = 4
```
Comment (chshersh) [inline src/Chapter3.hs]: If you derive the `Enum` typeclass for `House`, you can implement this function in a more concise way 😉

File: src/Chapter3.hs [hunk 5]
```diff
@@ -562,22 +642,30 @@ introducing extra newtypes.
...
 
-calculatePlayerDefense :: Int -> Int -> Int
-calculatePlayerDefense armor dexterity = armor * dexterity
+calculatePlayerDefense :: Armor -> Dexterity -> Defense
+calculatePlayerDefense (Armor a) (Dexterity d) = Defense $ a * d
 
-calculatePlayerHit :: Int -> Int -> Int -> Int
-calculatePlayerHit damage defense health = health + defense - damage
+calculatePlayerHit :: Damage -> Defense -> Health -> Health
+calculatePlayerHit (Damage dm) (Defense df) (Health h) = Health $ h + df - dm
```
Comment (chshersh) [inline src/Chapter3.hs:664]: Nice work on newtypes! :medal_sports: Newtypes are one very important feature in Haskell :+1:
You can read more about the newtype pattern in our blog post: https://kowainik.github.io/posts/haskell-mini-patterns#newtype

File: src/Chapter3.hs [hunk 6]
```diff
@@ -973,6 +1087,29 @@ implement the following functions:
...
+         | Saturday
+         deriving (Show, Eq, Enum, Ord)
+
+isWeekend :: Day  -> Bool
+isWeekend Saturday = True
+isWeekend Sunday   = True
+isWeekend _        = False
+
+nextDay :: Day -> Day
+nextDay d = cycle [Sunday .. Saturday] !! (fromEnum d + 1)
```
Comment (chshersh) [inline src/Chapter3.hs]: Okay, I see what this does. It's a clever solution! But it took me a while to understand what's happening here 😅 

One potential problem with this solution: if you decide to switch your data type to the European system and move `Sunday` in the end, it'll stop working. However, there's a way to implement this function that doesn't depend on which day of the week is last.

Another potential improvement: this function takes more steps to find the next day than it could. Ideally, it should take one or two steps. Maybe you can use a different function of the `Enum` typeclass to implement it 🤔

File: src/Chapter3.hs [hunk 7]
```diff
@@ -1008,7 +1145,175 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
-
+data KnightAction = DrinkPotion Health
+                  | CastSpell Defense
+                  deriving Show
+
+-- ugly empty type
+data MonsterAction = RunAway deriving Show
```
Comment (chshersh) [inline src/Chapter3.hs:1153]: There's nothing wrong with empty data types in general 🙂 
Having a separate data type allows you to extend it in the future if `Monster` starts having new actions.

File: src/Chapter3.hs [hunk 8]
```diff
@@ -1008,7 +1145,175 @@ Implement data types and typeclasses, describing such a battle between two
...
+data BattleAction = MonsterAc MonsterAction | KnightAc KnightAction deriving Show
+
+-- A name to help identify them
+newtype Name = Name String deriving Show
+
+-- The true information holder for the knight and monster
+data Entity = Entity { eName :: Name
+                     , eHealth :: Health
+                     , eAttack :: Attack
+                     , actions :: [Either BattleAction BattleAttack]
```
Comment (chshersh) [inline src/Chapter3.hs:1167]: One potential problem with this design is that it's possible for a monster to have a knight's action and vice versa. Ideally, we want to make illegal states unrepresentable and avoid such situations in total. So, for this, you correctly introduced separate data types `KnightAction` and `MonsterAction`. If you can extend them with other actions (e.g. `Attack`) and store only them in `Knight` and `Monster` correspondingly, you'll solve this problem.

This will solve this problem but will introduce a new challenge. From your code, I see that you probably tried that and experienced some difficulties. For instance, if you implement such a design, you won't be able to have a following function:

```haskell
getActions :: a -> [Either BattleAction BattleAttack]
```

You can have this one instead:

```haskell
getActions :: a -> [Either MonsterAction KnightAction]
```

But it will be awkward to work with. And you can see that it's not extendable if you decide to add new types of actions in the future.

One possible solution is to change the design of the `FighterAble` typeclass to something like this:

```haskell
class FighterAble me where
    performAction :: FighterAble other => me -> other -> (me, other)
    ... other needed methods ...
```

In this `performAction` you have access to the internals of the `me` type, so you can take its next action and perform it. But you don't have access to the internals of `other` and work with it using only other methods of the `FighterAble` typeclass.

So, this is just an idea of an alternative design 🙂 No need to implement this unless you want and have time! You did an amazing job solving this task and documenting your solution and it was a joy to read 🤗

**Review by chshersh — APPROVED** [review]
Excellent solutions to Chapter Three 🏆 

I left some comments and suggestions 🙂 I hope you can find them helpful

**Comment by aiwaverse** [comment]
I have updated the file with the improvements you suggested on the days question and the castle question, about the MultiParam extension, my first idea of a solution used it, but I hit a wall and couldn't progress with it, and forgot to delete de pragma.

And thank you for all the help and suggestions ^^

**Comment by chshersh** [comment]
@aiwaverse Your new updates look great! Feel free to merge it 🏁
